staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/promise/promise_test.go

```diff
@@ -17,105 +17,107 @@ limitations under the License.
 package promise
 
 import (
-	"sync"
-	"sync/atomic"
+	"context"
 	"testing"
 	"time"
 
-	testclock "k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/testing/eventclock"
+	"k8s.io/apimachinery/pkg/util/wait"
 )
 
-func TestLockingWriteOnce(t *testing.T) {
-	now := time.Now()
-	clock, counter := testclock.NewFake(now, 0, nil)
-	var lock sync.Mutex
-	wr := NewWriteOnce(&lock, counter)
-	var gots int32
-	var got atomic.Value
-	counter.Add(1)
+func TestWriteOnceSet(t *testing.T) {
+	oldTime := time.Now()
+	cval := &oldTime
+	ctx, cancel := context.WithCancel(context.Background())
+	wr := NewWriteOnce(nil, ctx.Done(), cval)
+	gots := make(chan interface{})
 	go func() {
-		lock.Lock()
-		defer lock.Unlock()
-
-		got.Store(wr.Get())
-		atomic.AddInt32(&gots, 1)
-		counter.Add(-1)
+		gots <- wr.Get()
 	}()
-	clock.Run(nil)
-	time.Sleep(time.Second)
-	if atomic.LoadInt32(&gots) != 0 {
+	select {
+	case <-gots:
 		t.Error("Get returned before Set")
+	case <-time.After(3 * time.Second):
```

Comment:
I suggest lowering to 1s.

In general, if it misbehaves, it should behave immediately. We make higher timeouts for cpu-starved environment to avoid flakiness, but here it's much better to shorten the test than have it artificially longer to switch from "flakiness to failing" in case of regression.

Reply:
I am not sure I understand what you are trying to say here.
If the implementation is correct, this test func will _never_ fail, regardless of what the timeout is and how loaded the CPU is.
If the implementation is incorrect, a longer timeout means the test will fail even for a more heavily loaded CPU.

Why is 3 seconds a problem?  This whole file now takes 12 seconds to run.

Reply:
12s is relatively long, especially given that our of those 12s, it's mostly sleeping. When I'm iterating with the code I would really like the tests to be as fast as possible.

Let me try to clarify what I meant:
- if the imlementation is correct - it will never fail
- if the implementation is incorrect (e.g. Get actually returns), then the longer timeout, the bigger probability of failure [it will probably not fail with 1us sleep in majority of cases, but will probably always fail with 1s]
What I'm saying is that in general, we make timeouts relatively long (waitTimeForever is 30s) to ensure that we won't get flakiness in heavy-loaded environments with huge cpu-starvation.
But that's not the case here - here you put 3s but it doesn't add any more value than what 1s would do - bigger timeouts won't help for potential flakiness.
In such case, I'm against artifically slowing down the tests, because that slows development and doesn't provide any real benefit.

Reply:
The timeout matters even without regression.
If the test subject is incorrect then an overloaded CPU can cause this test func to get the wrong answer.
If the test subject is correct then an overloaded CPU can cause this test func to get the wrong answer (we are not guaranteed that the happy path will complete before the timeout path).
A longer timeout makes it less likely (i.e., require more extreme circumstances) for this test func to get the wrong answer --- both for a correct and for an incorrect test subject.

This test file now takes 12 seconds to run.  Is that a problem?